### PR TITLE
Implementation of FastTimer

### PR DESCRIPTION
**Pull request by peterbryntesson**
## Purpose
This PR introduces the new FastTimer class and it's use for when we want really high-frequency changes for fast nodes.

## Does this introduce a breaking change?
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
```
Run the opc-plc with --vfr=49 or below

* Test the code

```
* Run the test suite
```

### CODE REVIEW

File: src/PlcNodeManager.cs [hunk 1]
```diff
@@ -92,6 +92,22 @@ public void UpdateFastNodes(object state, ElapsedEventArgs elapsedEventArgs)
             }
         }
 
+#pragma warning disable IDE0060 // Remove unused parameter
+        public void UpdateVeryFastNodes(object state, FastTimerElapsedEventArgs elapsedEventArgs)
```
Comment (buzzfrog) [inline src/PlcNodeManager.cs:96]: Do we really need to methods with the same code?

Comment (peterbryntesson) [reply src/PlcNodeManager.cs:96]: Well, we need two outer methods - but may then can call the same internal function. I'll refactor.

File: src/TimeService.cs [hunk 2]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+            nextTrigger += intervalLocal;
+
+            while (true)
+            {
+                var elapsed = sw.ElapsedTicks * tickFrequency;
+                var diff = nextTrigger - elapsed;
+                if (diff <= 0f)
+                    break;
+
+                if (diff < 1f)
```
Comment (buzzfrog) [inline src/TimeService.cs:230]: Is our coding standard in this project that we don't need {} for one liners?

Comment (peterbryntesson) [reply src/TimeService.cs:230]: Yes

**Comment by buzzfrog** [comment]
LGTM

**Comment by buzzfrog** [comment]
LGTM

**Review by buzzfrog — APPROVED** [review]

File: src/TimeService.cs [hunk 3]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+        /// The event handler we call when the timer is triggered
+        /// </summary>
+        public event FastTimerElapsedEventHandler Elapsed;
+
+        /// <summary>
+        /// Starts the timer 
+        /// </summary>
+        public void Start()
+        {
+            if (!isRunning)
```
Comment (simonjaeger) [inline src/TimeService.cs:169]: Is there a potential problem if two threads access the timer and try to start it? Potential race condition?

**Review by simonjaeger — APPROVED** [review]
Really cool! LGTM

File: src/PlcSimulation.cs [hunk 4]
```diff
@@ -103,7 +103,10 @@ public void Start()
 
             if (FastNodeCount > 0)
             {
-                _fastNodeGenerator = TimeService.NewTimer(_plcServer.PlcNodeManager.UpdateFastNodes, FastNodeRate);
+                // only use the fast timers when we need to go really fast
+                _fastNodeGenerator = FastNodeRate >= 50 ?
```
Comment (simonjaeger) [inline src/PlcSimulation.cs:107]: Great!

File: src/TimeService.cs (hunk 3 above)
Comment (peterbryntesson) [reply src/TimeService.cs:169]: Yes, I was waiting for someone to spot this... I'll guard this. Must run now but I'll fix it tonight.

File: src/PlcSimulation.cs [hunk 5]
```diff
@@ -103,7 +103,10 @@ public void Start()
 
             if (FastNodeCount > 0)
             {
-                _fastNodeGenerator = TimeService.NewTimer(_plcServer.PlcNodeManager.UpdateFastNodes, FastNodeRate);
+                // only use the fast timers when we need to go really fast
```
Comment (algattik) [inline src/PlcSimulation.cs:106]: maybe explain the reason (i.e. the FastTimer uses more resources)

Comment (algattik) [reply src/PlcSimulation.cs:106]: nit: should we consider using only FastTimers? That might simplify the codebase

File: src/TimeService.cs [hunk 6]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
         
         double Interval { get; set; }
     }
+
+    public class FastTimerElapsedEventArgs : EventArgs 
+    {
+        public FastTimerElapsedEventArgs()
```
Comment (algattik) [inline src/TimeService.cs:100]: delete?

File: src/TimeService.cs [hunk 7]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+        public FastTimer()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the FastTimer class and sets all properties to their default values except interval
+        /// </summary>
+        /// <param name="interval"></param>
+        public FastTimer(double interval)
+            : base()
```
Comment (algattik) [inline src/TimeService.cs:121]:
```suggestion
```

File: src/TimeService.cs [hunk 8]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+                    else
+                    {
+                        Thread.SpinWait(1000);
+                        Thread.Sleep(0);
+                    }
+
+                    // if we have a larger time to wait, we check if the interval has been changed in the meantime
+                    var newInterval = Interval;
+
+                    if (intervalLocal != newInterval)
```
Comment (algattik) [inline src/TimeService.cs:247]: Equality comparison of floating point numbers. Possible loss of precision while rounding values

File: src/TimeService.cs [hunk 9]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+                    return;
+            }
+        }
+
+        public void Dispose()
+        {
+            isRunning = false;
+        }
+
+        private static readonly float tickFrequency = 1000f / Stopwatch.Frequency;
```
Comment (algattik) [inline src/TimeService.cs:264]:
```suggestion
        private static readonly float TickFrequency = 1000f / Stopwatch.Frequency;
```

File: src/TimeService.cs [hunk 10]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+
+        private bool isEnabled = true;
```
Comment (algattik) [inline src/TimeService.cs:266]:
```suggestion
        private bool _isEnabled = true;
```
also for next line

File: src/TimeService.cs [hunk 11]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+        public bool AutoReset { get; set; }
+
+        /// <summary>
+        /// Is this timer currently running?
+        /// </summary>
+        public bool Enabled
+        {
+            get
+            {
+                return this.isEnabled;
```
Comment (algattik) [inline src/TimeService.cs:138]: unnecessary uses of `this`

Comment (algattik) [inline src/TimeService.cs:138]:
```suggestion
                get => this.isEnabled;
```

File: src/TimeService.cs [hunk 12]
```diff
@@ -68,4 +94,176 @@ public interface ITimer : IDisposable
...
+
+        /// <summary>
+        /// The event handler we call when the timer is triggered
+        /// </summary>
+        public event FastTimerElapsedEventHandler Elapsed;
+
+        /// <summary>
+        /// Starts the timer 
+        /// </summary>
+        public void Start()
```
Comment (algattik) [inline src/TimeService.cs:167]: why are these methods public? Since we have coverage only for their private use, they should probably be private

**Review by algattik — CHANGES_REQUESTED** [review]
Test fails on my machine:

OpcPlc.Tests.FastTimerTests.FastTimer_ShouldFire_100TimesPerSecond

Expected _callbacks.Count to be between 199 and 201, but found 1.

File: src/PlcSimulation.cs (hunk 5 above)
Comment (peterbryntesson) [reply src/PlcSimulation.cs:106]: Yeah, maybe we should only use FastTimers. It's worth thinking about

**Comment by peterbryntesson** [comment]
> Test fails on my machine:
> 
> OpcPlc.Tests.FastTimerTests.FastTimer_ShouldFire_100TimesPerSecond
> 
> Expected _callbacks.Count to be between 199 and 201, but found 1.

Fixed

File: src/PlcSimulation.cs (hunk 5 above)
Comment (peterbryntesson) [reply src/PlcSimulation.cs:106]: Nah, expanded on the comment instead.

File: src/TimeService.cs (hunk 8 above)
Comment (peterbryntesson) [reply src/TimeService.cs:247]: If that happens it will only adjust the interval slightly.

**Comment by peterbryntesson** [comment]
@algattik - addressed all your concerns. Mind giving it another look an approve if you think it's good?

**Review by algattik — APPROVED** [review]

**Pull request merged**